### ISSUE TITLE
docker compose no module name 'torch'

### ISSUE THREAD

**Issue by airdogvan** [oobabooga/text-generation-webui#4176]
### Describe the bug

Ubuntu 22.04 LTS
After running docker compose up --build 

got

  × python setup.py egg_info did not run successfully.
#9 11.09   │ exit code: 1
#9 11.09   ╰─> [6 lines of output]
#9 11.09       Traceback (most recent call last):
#9 11.09         File "<string>", line 2, in <module>
#9 11.09         File "<pip-setuptools-caller>", line 34, in <module>
#9 11.09         File "/tmp/pip-install-g3b1gyct/xformers_ece11b88871a456581ec13672a75669e/setup.py", line 18, in <module>
#9 11.09           import torch
#9 11.09       ModuleNotFoundError: No module named 'torch'
#9 11.09       [end of output]

Any help appreciated.

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Reproduction

docker compose up --build

### Logs

```shell
× python setup.py egg_info did not run successfully.
#9 11.09   │ exit code: 1
#9 11.09   ╰─> [6 lines of output]
#9 11.09       Traceback (most recent call last):
#9 11.09         File "<string>", line 2, in <module>
#9 11.09         File "<pip-setuptools-caller>", line 34, in <module>
#9 11.09         File "/tmp/pip-install-g3b1gyct/xformers_ece11b88871a456581ec13672a75669e/setup.py", line 18, in <module>
#9 11.09           import torch
#9 11.09       ModuleNotFoundError: No module named 'torch'
#9 11.09       [end of output]
```

### System Info

```shell
Ubuntu 22.04 LTS
```

**Comment by msmania** [comment]
I got the same error too.  It's never happened until this morning.

My solution was to replace the following line in `Dockerfile`

```
RUN --mount=type=cache,target=/root/.cache/pip,rw \
    . /app/venv/bin/activate && \
    pip3 install --upgrade pip setuptools wheel && \
    pip3 install torch torchvision torchaudio sentence_transformers xformers
```

with this one.

```
RUN --mount=type=cache,target=/root/.cache/pip,rw \
    . /app/venv/bin/activate && \
    pip3 install --upgrade pip setuptools wheel && \
    pip3 install torch==2.1.0 torchvision torchaudio sentence_transformers && \
    pip3 install xformers
```

Hopefully it works for you.

**Comment by airdogvan** [comment]
@msmania thanks so much, that worked fine.  Turns out I don't have the right CUDA version (11.5 instead of 11.8) but that's another problem.
Again many thanks for your help!!!

**Comment by swthorn** [comment]
Also having this issue. @airdogvan Did you happen to resolve the CUDA version mismatch after applying msmania's change to the dockerfile? I tried doing the following to target CUDA 11.8 (from the pytorch website), but I get the same error.

```
RUN --mount=type=cache,target=/root/.cache/pip,rw \
    . /app/venv/bin/activate && \
    pip3 install --upgrade pip setuptools wheel && \
    pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118 && \
    pip3 install sentence_transformers && \
    pip3 install xformers
```

**Comment by mongolu** [comment]
Why not using one-click included in docker file?

**Comment by airdogvan** [comment]
In fact I misread the error.  The real error is this one:

RuntimeError: 
#26 1.475 The detected CUDA version (11.8) mismatches the version that was used to compile
#26 1.475 PyTorch (12.1). Please make sure to use the same CUDA versions.

Not very sure what that means...

On my host I have version 11.5, not sure how or if that impacts the docker environment...

**Comment by Reezlaw** [comment]
> In fact I misread the error. The real error is this one:
> 
> RuntimeError: #26 1.475 The detected CUDA version (11.8) mismatches the version that was used to compile #26 1.475 PyTorch (12.1). Please make sure to use the same CUDA versions.
> 
> Not very sure what that means...
> 
> On my host I have version 11.5, not sure how or if that impacts the docker environment...

I found that torch==2.1.0 is a mismatch with the CUDA version, use 2.0.1

**Comment by airdogvan** [comment]
@Reezlaw thanks for the help, but no go.  Restarted from scratch with your suggestion and still the same error... :(

**Comment by jFkd1** [comment]
> @Reezlaw thanks for the help, but no go. Restarted from scratch with your suggestion and still the same error... :(

Here's my fix: replace any line that installs torch in dockerfile with: 
```
pip3 install torch==2.0.1+cu118 torchvision==0.15.2+cu118 torchaudio==2.0.2 --index-url https://download.pytorch.org/whl/cu118
```

**Comment by bannsec** [comment]
> > @Reezlaw thanks for the help, but no go. Restarted from scratch with your suggestion and still the same error... :(
> 
> Here's my fix: replace any line that installs torch in dockerfile with:
> 
> ```
> pip3 install torch==2.0.1+cu118 torchvision==0.15.2+cu118 torchaudio==2.0.2 --index-url https://download.pytorch.org/whl/cu118
> ```

Having same issue. This caused the build to finish but upon run it gives an error:

```
Make sure to replace username:password with your own.
text-generation-webui-text-generation-webui-1  | Traceback (most recent call last):
text-generation-webui-text-generation-webui-1  |   File "/app/server.py", line 30, in <module>
text-generation-webui-text-generation-webui-1  |     from modules import (
text-generation-webui-text-generation-webui-1  |   File "/app/modules/chat.py", line 18, in <module>
text-generation-webui-text-generation-webui-1  |     from modules.text_generation import (
text-generation-webui-text-generation-webui-1  |   File "/app/modules/text_generation.py", line 24, in <module>
text-generation-webui-text-generation-webui-1  |     from modules.models import clear_torch_cache, local_rank
text-generation-webui-text-generation-webui-1  |   File "/app/modules/models.py", line 22, in <module>
text-generation-webui-text-generation-webui-1  |     from modules import RoPE, llama_attn_hijack, sampler_hijack
text-generation-webui-text-generation-webui-1  |   File "/app/modules/llama_attn_hijack.py", line 7, in <module>
text-generation-webui-text-generation-webui-1  |     import transformers.models.llama.modeling_llama
text-generation-webui-text-generation-webui-1  |   File "/app/venv/lib/python3.10/site-packages/transformers/models/llama/modeling_llama.py", line 45, in <module>
text-generation-webui-text-generation-webui-1  |     from flash_attn import flash_attn_func, flash_attn_varlen_func
text-generation-webui-text-generation-webui-1  |   File "/app/venv/lib/python3.10/site-packages/flash_attn/__init__.py", line 3, in <module>
text-generation-webui-text-generation-webui-1  |     from flash_attn.flash_attn_interface import (
text-generation-webui-text-generation-webui-1  |   File "/app/venv/lib/python3.10/site-packages/flash_attn/flash_attn_interface.py", line 8, in <module>
text-generation-webui-text-generation-webui-1  |     import flash_attn_2_cuda as flash_attn_cuda
text-generation-webui-text-generation-webui-1  | ImportError: /app/venv/lib/python3.10/site-packages/flash_attn_2_cuda.cpython-310-x86_64-linux-gnu.so: undefined symbol: _ZN3c104cuda9SetDeviceEi
```

**Comment by sammcj** [comment]
Same issue here.

I know this project has no tests / testing framework but I think it should at least have a build that runs on PR creation that has to pass before the PR can be merged. This way we could be sure that dependencies like this don't become broken.

**Comment by sammcj** [comment]
I managed to fix it by updating the container base image to the current CUDA version (12.2) - https://github.com/oobabooga/text-generation-webui/pull/4256

**Comment by airdogvan** [comment]
can we all know how you replaced the base image?

I tried FROM nvidia/cuda:12.2-devel-ubuntu22.04 as builder 
but got as result:
failed to solve: nvidia/cuda:12.2-devel-ubuntu22.04: docker.io/nvidia/cuda:12.2-devel-ubuntu22.04: not found

**Comment by sammcj** [comment]
You’re missing the 0.

https://hub.docker.com/layers/nvidia/cuda/12.2.0-devel-ubuntu22.04/

**Comment by airdogvan** [comment]
thanks for the precision.  That apparently did build correctly but I now seem to have a system level problem:  Error response from daemon: could not select device driver "nvidia" with capabilities: [[gpu]]

As already said, on my host I have 11.5 installed and I suppose the container has, at one time or another, to interact with the physical driver.  The versions being different (12.2 versus 11.5) causing a conflict...
Not sure if there's a way for me to install 12.2.  Let me do some research...

But again thanks for the correction...

**Comment by sammcj** [comment]
Do you mean 11.8? 11.5 is quite old / not supported.

**Comment by airdogvan** [comment]
no was 11.5.  Anyway uninstalled that one and installed 12.2 and got exactly the same error... ???  Not sure what to do now...  Cuda is installed and working, when I do 
watch -d -n 0.5 nvidia-smi

I get the correct information telling me how much memory is used and other info...   So cuda is there and working

**Comment by sammcj** [comment]
Turns out it wasn't so much CUDA 12.2 itself that fixed it in my PR, but installing a different version of torch (which happened to be 12.1 / 12.2).

https://github.com/oobabooga/text-generation-webui/pull/4256/files#diff-f34da55ca08f1a30591d8b0b3e885bcc678537b2a9a4aadea4f190806b374ddcR46

---

Regardless - because this wasn't merged (due to potential backwards compatibility issues with newer CUDA) - text-generation-webui remains broken.

**Comment by airdogvan** [comment]
Sorry, forgot to let you know that it's now working for me.   Turns out I needed to install the nvidia docker package. 

It works but the output is incredibly slow, about 2 words/second, which is weird.  I ran AI models on the CPU which results in the model " thinking"  a good while before starting to answer.   But on GPU it starts to answer immediately but I have to be very patient for the rest. Maybe there's a parameter that I need to tune correctly???  Don't know
